BlueBrain/CoreNeuron
# Define NrnCoreTransferEvents in global namespace.

# Description
This PR contains fixes for issues uncovered by the UndefinedBehaviour sanitizer: https://clang.llvm.org/docs/UndefinedBehaviorSanitizer.html

##  Define NrnCoreTransferEvents in global namespace.
A pointer to this type is passed across the NEURON-CoreNEURON boundary
by the `nrn2core_transfer_tqueue callback`, and it is undefined behaviour
to call a function via a pointer to a different function type. This is
what was happening before due to a mismatch of `NrnCoreTransferEvents` and
`coreneuron::NrnCoreTransferEvents`:

```
82: {path}/nrn/external/coreneuron/coreneuron/io/nrn2core_data_init.cpp:152:39: runtime error: call to function nrn2core_transfer_tqueue through pointer to incorrect function type 'coreneuron::NrnCoreTransferEvents *(*)(int)'
82: nrncore_callbacks.cpp:741: note: nrn2core_transfer_tqueue defined here
82:     #0 0x10eb240ea in coreneuron::nrn2core_tqueue() nrn2core_data_init.cpp:152
82:     #1 0x10eaea2b6 in coreneuron::nrn_init_and_load_data(int, char**, coreneuron::CheckPoints&, bool, bool) main1.cpp:334
82:     #2 0x10eaeb986 in run_solve_core main1.cpp:508
82:     #3 0x10bc01554 in corenrn_embedded_run enginemech.cpp:108
82:     #4 0x10a12487f in nrncore_run(char const*) nrncore_write.cpp:303
82:     #5 0x10a124e97 in nrncore_psolve(double, int) nrncore_write.cpp:349
82:     #6 0x10a1c292c in psolve(void*) ocbbs.cpp:671
82:     #7 0x10a385a46 in hoc_call_ob_proc hoc_oop.cpp:759
82:     #8 0x10a388c1d in hoc_object_component() hoc_oop.cpp:1174
82:     #9 0x10a346c41 in hoc_execute(Inst*) code.cpp:2682
82:     #10 0x10a34f033 in hoc_ifcode() code.cpp
82:     #11 0x10a346c41 in hoc_execute(Inst*) code.cpp:2682
82:     #12 0x10a351356 in hoc_call() code.cpp:1416
82:     #13 0x10a346c41 in hoc_execute(Inst*) code.cpp:2682
82:     #14 0x10a37a0dc in hoc_run1() hoc.cpp:1410
82:     #15 0x10a379744 in hoc_main1(int, char const**, char const**) hoc.cpp:1054
82:     #16 0x109cd20cd in ivocmain_session(int, char const**, char const**, int) ivocmain.cpp:839
82:     #17 0x109cd0e75 in ivocmain(int, char const**, char const**) ivocmain.cpp:415
82:     #18 0x109bc6f55 in main nrnmain.cpp:53
82:     #19 0x7fff20423f3c in start+0x0 (libdyld.dylib:x86_64+0x15f3c)
```

**How to test this?**
```bash
cmake .. -G Ninja \
  -DCMAKE_C_FLAGS="-fsanitize=undefined -fno-omit-frame-pointer -g" \
  -DCMAKE_CXX_FLAGS="-fsanitize=undefined -fno-omit-frame-pointer -g" \
  -DNRN_ENABLE_CORENEURON=ON \
  -DCORENRN_ENABLE_NMODL=ON \
  -DNRN_ENABLE_TESTS=ON \
  -DNRN_ENABLE_INTERVIEWS=OFF \
  -DNRN_ENABLE_TESTS_BBP=ON \
  -DNRN_CLANG_FORMAT=ON \
  -DNRN_CMAKE_FORMAT=ON \
  -DNRN_FORMATTING_ON="since-ref:master" \
  -DNRN_FORMATTING_CPP_CHANGES_ONLY=ON
cmake --build . --parallel
export 'UBSAN_OPTIONS=print_stacktrace=1'
ctest -V
```
Check the test logfile for messages from the sanitizer.

**Test System**
 - OS: macOS 11.6
 - Compiler: AppleClang 13
 - Version: master
 - Backend: CPU

**Use certain branches for the SimulationStack CI**

CI_BRANCHES:NEURON_BRANCH=master,
